components/fxa-client/src/oauth.rs

```diff
@@ -172,25 +172,47 @@ impl FirefoxAccount {
         &self,
         auth_params: AuthorizationParameters,
     ) -> Result<String> {
-        // TODO: Add validation to check for which keys the client is allowed to request,
-        //       And if we have those keys stored locally.
-        //       Currently we pull ALL available keys that match the requested scopes, regardless
-        //       if the client is actually allowed to request them.
         let session_token = self.get_session_token()?;
+
+        // Validate request to ensure that the client is actually allowed to request
+        // the scopes they requested
+        let valid_scopes = self.client.scoped_key_data(
```

Comment:
Nit: this is begging to be called `allowed_scopes` rather than `valid_scopes`, both because it's more precise and because the surrounding comments etc use the word "allowed" :-)